pkg/storage/s3.go

```diff
@@ -1,56 +1,318 @@
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/defaults"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/spf13/pflag"
+
 	"github.com/pingcap/errors"
 	"github.com/pingcap/kvproto/pkg/backup"
-	"github.com/spf13/pflag"
 )
 
 const (
-	s3EndpointOption = "s3.endpoint"
-	s3RegionOption   = "s3.region"
+	s3EndpointOption     = "s3.endpoint"
+	s3RegionOption       = "s3.region"
+	s3StorageClassOption = "s3.storage-class"
+	s3SSEOption          = "s3.sse"
+	s3ACLOption          = "s3.acl"
+	s3ProviderOption     = "s3.provider"
+	// accessKeyEnv         = "AWS_ACCESS_KEY_ID"
+	// secretAccessKeyEnv   = "AWS_SECRET_ACCESS_KEY"
+	notFound = "NotFound"
+	// number of retries to make of operations
+	maxRetries = 3
 )
 
-// S3BackendOptions are options for configurating the S3 storage.
+// s3Handlers make it easy to inject test functions
+type s3Handlers interface {
+	HeadObject(*s3.HeadObjectInput) (*s3.HeadObjectOutput, error)
+	GetObject(*s3.GetObjectInput) (*s3.GetObjectOutput, error)
+	PutObject(*s3.PutObjectInput) (*s3.PutObjectOutput, error)
+	HeadBucket(*s3.HeadBucketInput) (*s3.HeadBucketOutput, error)
+	WaitUntilObjectExists(*s3.HeadObjectInput) error
+}
+
+// S3Storage info for s3 storage
+type S3Storage struct {
+	session *session.Session
+	// svc     *s3.S3
+	svc     s3Handlers
+	options *backup.S3
+}
+
+// S3BackendOptions contains options for s3 storage
 type S3BackendOptions struct {
-	Endpoint string `json:"endpoint" toml:"endpoint"`
-	Region   string `json:"region" toml:"region"`
+	Endpoint              string `json:"endpoint" toml:"endpoint"`
+	Region                string `json:"region" toml:"region"`
+	StorageClass          string `json:"storage_class" toml:"storage_class"`
+	SSE                   string `json:"sse" toml:"sse"`
+	ACL                   string `json:"acl" toml:"acl"`
+	AccessKey             string `json:"access_key" toml:"access_key"`
+	SecretAccessKey       string `json:"secret_access_key" toml:"secret_access_key"`
+	Provider              string `json:"provider" toml:"provider"`
+	ForcePathStyle        bool   `json:"force_path_style" toml:"force_path_style"`
+	UseAccelerateEndpoint bool   `json:"use_accelerate_endpoint" toml:"use_accelerate_endpoint"`
 }
 
 func (options *S3BackendOptions) apply(s3 *backup.S3) error {
 	if options.Endpoint == "" && options.Region == "" {
 		return errors.New("must provide either 's3.region' or 's3.endpoint'")
 	}
-	// TODO: Verify Region.
+	if options.Endpoint != "" {
+		if !strings.HasPrefix(options.Endpoint, "https://") &&
+			!strings.HasPrefix(options.Endpoint, "http://") {
+			options.Endpoint = "http://" + options.Endpoint
+		}
+	}
+	if options.Provider == "alibaba" || options.Provider == "netease" ||
```

Comment:
Could you add a document to this for why this is needed? For example, the documentation link to it.

Reply:
done